### DISCUSSION TITLE
How to load a shared library

### DISCUSSION THREAD

**Discussion by Angel-del-dev** [vlang/v#22147] [Questions and Answers] [answered]
Vlang version : `V 0.4.7 fc31b44`

I have the shared libraries in 'lib/'.

The library I'd like to load is 'functions.so' and it contains only one function named 'fib'.

I haven't been able to use this library with the information in the [docs](https://modules.vlang.io/dl.loader.html)

The Shared library was compiled with V too. 
`v -shared -o lib/functions export/functions.v`
```
// export/functions.v
module export

@[export: 'fib']
pub fn fib(n int) int {
	if n <= 2 { return 1 }
	else { return fib(n-1) + fib(n-2) }
}
```

Loading the library:
```
module main

import os
import dl.loader

fn main() {
	abs_path := os.abs_path('lib/functions.so')

	mut dl_loader := loader.get_or_create_dynamic_lib_loader(
		key: 'Lib',
		paths: [abs_path]
	) or {
		panic(err)
	}
	print(dl_loader)
}
```

And compiling it with: `v main.v && ./main`

It outputs the following:

```
&loader.DynamicLibLoader{
    key: 'Lib'
    flags: 1
    paths: ['/home/myuser/Desktop/dev/project/v/dllgenerator/lib/functions.so']
    handle: 0
    sym_map: {}
}
```
The `sym_map`  is empty and I believe its because V doesn't find the shared library.

 I have tried using both relative and absolute paths.
The file exists, and i can read it with V

I also tried changing the module name in `functions.v` to see if it was giving some incompatibility with `@[export: 'fib']`.

File tree:

- export
   - functions.v
- lib
  - functions.so
- main.v

**Comment by ikolesnikes** [comment, marked as answer]
Is seems that get_or_create_dynamic_lib_loader doesn't load any symbols automatically. Probably, this is by design.
If you do
`sym := dl_loader.get_sym('fib')!`
then `sym_map` gets updated.

Looking at the `loader.v`, `sym_map` is being used as a cache for symbols resolved by `get_sym`.
